server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1016,15 +1066,103 @@ public void writeTo(StreamOutput out) throws IOException {
+            // are defined only for the default project or (b) no repositories at all. What we need to do are:
+            // 1. Iterate through the multi-project's MapDiff to extract the RepositoriesMetadata of the default project
+            // 2. Throws if any repositories are found for non-default projects
+            // 3. Merge default project's RepositoriesMetadata into Metadata#customs
+            final var combineClustersCustoms = new SetOnce<MapDiff<String, MetadataCustom, Map<String, MetadataCustom>>>();
+            final var updatedMultiProject = DiffableUtils.updateDiffsAndUpserts(multiProject, ignore -> true, (k, v) -> {
+                assert v instanceof ProjectMetadata.ProjectMetadataDiff : v;
+                final var projectMetadataDiff = (ProjectMetadata.ProjectMetadataDiff) v;
+                final var bwcCustoms = DiffableUtils.split(
+                    projectMetadataDiff.customs(),
+                    RepositoriesMetadata.TYPE::equals,
+                    PROJECT_CUSTOM_VALUE_SERIALIZER,
+                    type -> RepositoriesMetadata.TYPE.equals(type) == false,
+                    PROJECT_CUSTOM_VALUE_SERIALIZER
+                );
+                // Simply return if RepositoriesMetadata is not found
+                if (bwcCustoms.v1().isEmpty()) {
+                    return projectMetadataDiff;
+                }
+                // RepositoriesMetadata can only be defined for the default project. Otherwise throw exception.
+                if (ProjectId.DEFAULT.equals(k) == false) {
+                    throwForVersionBeforeRepositoriesMetadataMigration(out);
+                }
+                // RepositoriesMetadata is found for the default project as a diff, merge it into the Metadata#customs
+                combineClustersCustoms.set(
+                    DiffableUtils.<String, MetadataCustom, ClusterCustom, ProjectCustom, Map<String, MetadataCustom>>merge(
+                        clusterCustoms,
+                        bwcCustoms.v1(),
+                        DiffableUtils.getStringKeySerializer()
+                    )
+                );
+                return projectMetadataDiff.withCustoms(bwcCustoms.v2());
+            }, (k, v) -> {
+                final ProjectCustom projectCustom = v.customs().get(RepositoriesMetadata.TYPE);
+                // Simply return if RepositoriesMetadata is not found
+                if (projectCustom == null) {
+                    return v;
+                }
+                // RepositoriesMetadata can only be defined for the default project. Otherwise throw exception.
+                if (ProjectId.DEFAULT.equals(k) == false) {
+                    throwForVersionBeforeRepositoriesMetadataMigration(out);
                 }
+                // RepositoriesMetadata found for the default project as an upsert, package it as MapDiff and merge into Metadata#customs
+                combineClustersCustoms.set(
```

Comment:
I'm probably missing something here, but just to double check: is it guaranteed that only one of these lambdas reaches the end? In other words, are we sure we won't encounter non-empty repositories for the default project _in both upserts and diffs_? I just want to make sure `combineClusterCustoms.set()` is only called once.

Reply:
Yes, upserts and diffs are mutually exclusive, see 
https://github.com/elastic/elasticsearch/blob/dd58b0b6face632cd4476a496cb79e8c0dffdcb6/server/src/main/java/org/elasticsearch/cluster/DiffableUtils.java#L337-L348